esp/services/espcontrol/ws_espcontrolservice.cpp

```diff
@@ -51,3 +131,200 @@ bool CWSESPControlEx::onSetLogging(IEspContext& context, IEspSetLoggingRequest&
+    if (!globalLock)
+        throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Unable to connect to ESP Session information in dali %s", xpath);
+    return globalLock.getClear();
+}
+
+bool CWSESPControlEx::onSessionQuery(IEspContext& context, IEspSessionQueryRequest& req, IEspSessionQueryResponse& resp)
+{
+    try
+    {
+#ifdef _USE_OPENLDAP
+        CLdapSecManager* secmgr = dynamic_cast<CLdapSecManager*>(context.querySecManager());
+        if(secmgr && !secmgr->isSuperUser(context.queryUser()))
+            throw MakeStringException(ECLWATCH_SUPER_USER_ACCESS_DENIED, "Failed to query session. Permission denied.");
+#endif
+
+        StringBuffer fromIP = req.getFromIP();
+        CSessionState state = req.getState();
+        StringBuffer authDomain = req.getAuthDomain();
+        if (authDomain.isEmpty())
+            authDomain.set("default");
+        unsigned port = 8010;
+        if (!req.getPort_isNull())
+            port = req.getPort();
+
+        VStringBuffer xpath("/%s/%s[@name='%s']/%s[@name='%s']/%s[@port=\"%d\"]", PathSessionRoot, PathSessionProcess,
+            espProcess.get(), PathSessionDomain, authDomain.str(), PathSessionApplication, port);
+        Owned<IRemoteConnection> globalLock = querySDSConnection(xpath.str(), RTM_LOCK_READ, SESSION_SDS_LOCK_TIMEOUT);
+
+        IArrayOf<IEspSession> sessions;
+        if (!fromIP.trim().isEmpty())
+            xpath.setf("%s[%s='%s']", PathSessionSession, PropSessionNetworkAddress, fromIP.str());
+        else if (state != SessionState_Undefined)
+            xpath.setf("%s[%s='%d']", PathSessionSession, PropSessionState, state);
+        Owned<IPropertyTreeIterator> iter = globalLock->getElements(xpath.str());
+        ForEach(*iter)
+        {
+            IPropertyTree& sessionTree = iter->query();
+            Owned<IEspSession> s = createSession();
+            setSessionInfo(sessionTree.queryProp(PropSessionID), &sessionTree, authDomain.str(), port, s);
+            sessions.append(*s.getLink());
+        }
+        resp.setSessions(sessions);
+    }
+    catch(IException* e)
+    {
+        FORWARDEXCEPTION(context, e, ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWSESPControlEx::onSessionInfo(IEspContext& context, IEspSessionInfoRequest& req, IEspSessionInfoResponse& resp)
+{
+    try
+    {
+#ifdef _USE_OPENLDAP
+        CLdapSecManager* secmgr = dynamic_cast<CLdapSecManager*>(context.querySecManager());
+        if(secmgr && !secmgr->isSuperUser(context.queryUser()))
+            throw MakeStringException(ECLWATCH_SUPER_USER_ACCESS_DENIED, "Failed to get session information. Permission denied.");
+#endif
+
+        StringBuffer sessionID = req.getSessionID();
+        if (sessionID.trim().isEmpty())
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "SessionID not specified.");
+
+        StringBuffer authDomain = req.getAuthDomain();
```

Comment:
minor: StringAttr is more commonly used, where string manipulation is not needed (there a couple of other similar lines of code elsewhere)